src/System.CommandLine/Parsing/ArgumentResult.cs

```diff
@@ -55,115 +53,152 @@ public sealed class ArgumentResult : SymbolResult
+            CommandResult parent = (CommandResult)Parent!;
+            var arguments = parent.Command.Arguments;
+            int argumentIndex = arguments.IndexOf(Argument);
+            int nextArgumentIndex = argumentIndex + 1;
+            int tokensToPass = _tokens.Count - numberOfTokens;
 
-        internal ParseError? CustomError(Argument argument)
-        {
-            for (var i = 0; i < argument.Validators.Count; i++)
+            while (tokensToPass > 0 && nextArgumentIndex < arguments.Count)
             {
-                var validateSymbolResult = argument.Validators[i];
-                validateSymbolResult(this);
+                Argument nextArgument = parent.Command.Arguments[nextArgumentIndex];
+                ArgumentResult nextArgumentResult;
+
+                if (SymbolResultTree.TryGetValue(nextArgument, out SymbolResult? symbolResult))
+                {
+                    nextArgumentResult = (ArgumentResult)symbolResult;
+                }
+                else
+                {
+                    // it might have not been parsed yet or due too few arguments, so we add it now
+                    nextArgumentResult = new ArgumentResult(nextArgument, SymbolResultTree, Parent);
+                    SymbolResultTree.Add(nextArgument, nextArgumentResult);
+                }
 
-                if (!string.IsNullOrWhiteSpace(ErrorMessage))
+                while (!nextArgumentResult.IsArgumentLimitReached && tokensToPass > 0)
                 {
-                    return new ParseError(ErrorMessage!, Parent is OptionResult option ? option : this);
+                    Token toPass = _tokens[numberOfTokens];
+                    _tokens.RemoveAt(numberOfTokens);
+                    nextArgumentResult.AddToken(toPass);
+                    --tokensToPass;
                 }
+
+                nextArgumentIndex++;
+            }
+
+            // When_tokens_are_passed_on_by_custom_parser_on_last_argument_then_they_become_unmatched_tokens
+            while (tokensToPass > 0)
+            {
+                Token unmatched = _tokens[numberOfTokens];
+                _tokens.RemoveAt(numberOfTokens);
+                SymbolResultTree.AddUnmatchedToken(unmatched);
+                --tokensToPass;
             }
+        }
+
+        /// <inheritdoc/>
+        public override string ToString() => $"{GetType().Name} {Argument.Name}: {string.Join(" ", Tokens.Select(t => $"<{t.Value}>"))}";
 
-            return null;
+        /// <inheritdoc/>
+        public override void AddError(string errorMessage)
+        {
+            SymbolResultTree.AddError(new ParseError(errorMessage, Parent is OptionResult option ? option : this));
+            _conversionResult = ArgumentConversionResult.Failure(this, errorMessage, ArgumentConversionResultType.Failed);
         }
 
-        private ArgumentConversionResult Convert(Argument argument)
+        private ArgumentConversionResult ValidateAndConvert(bool potentialRecursion)
```

Comment:
this method is now responsible for entire validation. The problem is that it can be called by not only the `ParseResultVisitor` but also custom parsers and validator which can do things like `ArgumentResult.GetDefaultValue()`. For the latter case I've added `potentialRecursion` argument to ensure we don't have endless recursion (validator->getvalue->validator)

Reply:
The name `potentialRecursion` is a little unclear to me. Is this saying whether to allow recursion?